Make remote_data_updated mechanism usable for connectors

jgoizueta opened the pull request:
Connectors were missing access to the modified_at field of synchronizations.

For non-connector runner it worked because the modified_at date was passed through the downloader.

rafatower commented on services/importer/lib/importer/connector_runner.rb:
```diff
@@ -103,6 +108,7 @@ def checksum
       def last_modified
         # This method is needed to make the interface of ConnectorRunner compatible with Runner,
         # but we have no meaningful data to return here.
+        @connector.last_modified
```
Arg! by the spec below I deduce one is the remote modification time and the other the local modification time. :cry: 

I wonder if you could use aliases to clarify those :smile_cat:

jgoizueta replied:
Regarding the `modified_at` bug, I'm puzzled because I checked again during the rest of the demo and then it had the correct value.
Regarding the terrible modified times confusion, I'll try to find better names for the locals. In fact I did in the last tests I added and I called them `date_the_data_was_modified` (for `last_modified`) and `date_the_data_was_last_copied` (for `modified_at`)

jgoizueta replied:
Maybe `copied_at` (or `copied_into_carto_at`) for `modified_at` and `changed_at` (`source_changed_at`) for `last_modified` ?

rafatower replied:
`carto_last_synced`  and `remote_source_last_modified` ? I like to be explicit about what entity (carto or the remote source) the attribute qualifies. In fact I think it is the most important thing to seek with the naming in this case.

jgoizueta replied:
I think those are more clear, so let it be:
* `last_modified` -> `remote_source_last_modified`
* `modified_at` -> `carto_last_synced`
We have to live with the synchronizations metadata column being `modified_at` but I'll review to make it consistent in the code

jgoizueta replied:
I think it's worth renaming the connector `modified_at` parameter

jgoizueta replied:
I took a look, and the naming comes from `Synchronization::Member` && `Runner` which have a `last_modified` method which I kept for compatibility in `ConnectorRunner`(and then introduced in the Connectors/Providers)

The cherry on top is this line where the synchronizations `modified_at` is pass to the Downloader http_options as `last_modified`:
https://github.com/CartoDB/cartodb/blob/3de761f3bb379bf630fcd26622c55f47939c588b/app/models/synchronization/member.rb#L350

So now, I'm not sure how far to take this renaming
* I'll change local variables in this PR and db-connectors
* I'll change `modified_at` parameter recently introduced in the connector providers
* Should I change the `last_modified` method in Providers? The name should be kept in ConnectorRunner for compatibility with Runner, unless...
* I also rename the Runner and Adapter `last_modified` methods

jgoizueta replied:
Hey, sorry to add so much confusion to this confusing matter, but I've realised I'm not very sharp today:

The naming we were proposing may not be accurate.

last_modified is the date the source data was last modified, so I think `remote_source_last_modified` is ok.

but modified_at is not the date the data was synced or copied to carto, it is the date the remote source had last been modified when we last synced it into carto. 

Note how we update it:
https://github.com/CartoDB/cartodb/blob/3de761f3bb379bf630fcd26622c55f47939c588b/app/models/synchronization/member.rb#L392

So it could be `previous_remote_source_last_modified` or `remote_source_last_modified_at_last_sync`or something like that to be accurate. But that's probably overkill and not easy to grasp.

I'll give it a little more thought.

rafatower replied:
> but modified_at is not the date the data was synced or copied to carto, it is the date the remote source had last been modified when we last synced it into carto

then I'd say the problem is *not the naming* but that we're lacking something and instead have something else, similar but not quite, with a misleading name that made us think of the thing we needed.... I don't know if I'm explaining this correctly or am right or adding more confusion... o_O

rafatower replied:
better refer to the new comments, closing this thread of confusion :D

jgoizueta commented:
# Parameter-naming confusion

I've kept the `last_modified` name used pervasively through the code as the remote source update time, but I've changed the newly introduced `modified_at` connectors parameter to `previous_last_modified` to indicate that that is the external source last-modified date corresponding to the previous update (sync).

### So, how was the change detection originally implemented for file/url sources?

The `Runner` class, used by both imports and syncs, implements a `remote_data_updated?` method which calls `Downloader#modified?` which compares the last-modified date of the new data (from HTTP header) to the value from the previous sync, which is stored in `synchronizations.modified_at` and passed to the Downloader through `http_options.last_modified`

The `synchronizations.modified_at` column is set at the end of syncs through `set_success_state_from` from the importer's (a `Synchronization::Adapter` for syncs) `last_modified` method that takes the value from the `Runner` and this in turn from the `Downloader`.

DatasourceDownloader has similar a mechanism (modified?) using checksums instead of dates. Note that table `synchronizations` also has a `checksum` column to store this too.

Now, I see a problem with this mechanism, which is that the first import doesn't update the synchronization `modified_at`.
So the first sync will always be executed even if unnecessary. I'll open a ticket to fix this which should be a matter of adding `synchronization.modified_at = importer.last_modified` to `::DataImport::update_synchronization`.

### And how has this been extended to db connectors (and fixed)?

The `remote_data_updated?` has been implemented in `ConnectorRunner`, where it calls the connector provider's instance of this method. The provider should be able to implement this method because it receives through its constructor a `previous_last_import` parameter, that is taken from `synchronizations.modified_at`.

The ConnectorRunner also implements a `last_modified` modified method, also relayed to the provider, which allows the update of `synchronizations.modified_at` from the last modified data of the data just copied.

jgoizueta commented:
# :warning: This can break existing connectors :warning:

Since this introduces a new parameter, `previous_last_modified` unexpected by the connector providers, this must be deployed with the corresponding db-connectors changes. See the [clubhouse story](https://app.clubhouse.io/cartoteam/story/88832) for details
